ompi/datatype/ompi_datatype_create_indexed.c

```diff
@@ -31,13 +31,12 @@
 
 
 /* We try to merge together data that are contiguous */
-int32_t ompi_datatype_create_indexed( int count, const int* pBlockLength, const int* pDisp,
```

Comment:
Making the `count` in indexed datatypes a size_t makes very little sense, because it indicates the number of entries in the displacement and blocklen arrays. In other terms the user-facing datatype representation will be several gigabytes long.

I can understand that for symmetry with the contiguous one would like to have this as a size_t, but my comment above remains valid.

Reply:
@bosilca I need your thoughts here. I figured `MPI_Type_indexed` also has the large-count variant `MPI_Type_indexed_c`, which I imagine will also need `size_t` support in `ompi_datatype_create_indexed` and related functions. Did I missing anything?

Same goes for *struct and *vector.

Reply:
@wenduwan Check out the work that @hppritcha and Jakob did on the *w collectives. There are now disp and count arrays that store whether the source is 32 or 64bit and hand out 64bit consistently: https://github.com/open-mpi/ompi/blob/main/ompi/util/count_disp_array.h

Reply:
The count is the number of elements in the indexed type, and if that number cannot be represented as an int, it basically means that the datatype description (where we need 64 bytes per contiguous element to represent) will be bigger than the represented data. No normal person should even imagine such an API.

@devreal suggestion is mostly irrelevant here. In the collective case the user-provided buffers are guaranteed to be available during the entire collective operation, so piggybacking into the user buffer is possible. There is no such guarantees with the datatype.

Reply:
@bosilca Thanks. Could you elaborate more on "the datatype description .. will be bigger than the represented data." part? Are you referring to this struct? I noted that `opal_datatype_count_t` is already `size_t`
```
struct dt_type_desc_t {
    opal_datatype_count_t length; /**< the maximum number of elements in the description array */
    opal_datatype_count_t used;   /**< the number of used elements in the description array */
    dt_elem_desc_t *desc;
};
```

Reply:
Each data in the indexed/struct will be represented by an `ddt_elem_desc` struct which is 32 bytes. So, if you have a number of entries that will not fit into an int, it means what the data description itself will be over 64GB, there will be little memory left for the data itself (especially that if you use an index or struct you have gaps between these elements).

Reply:
The vector representation is compact because its regular and repetitive, with a single `ddt_elem_desc` (aka. 64 bytes) one can describe a datatype that covers the entire memory. Indexed and struct representations have a number of `ddt_elem_desc` entries (I'm not talking about the datatype count here), and when the number of entries is larger than an int (which is what we are talking about here), the datatype representation itself will be comparable with the memory layout it covers.

Reply:
Ah I see. `ompi_datatype_create_vector` is not calling `ompi_datatype_create` in a loop, but indexed/struct does.

But still, without embiggening the indexed/struct count(as well as block length, displ) how can we support the `MPI_Type_{indexed,struct}_c` APIs?

Reply:
My comment was mostly about how stupid that API is, not about how we support it. And I'm not even talking about the performance of parsing that extremely large description to pack/unpack data.

But if you ask me how to do it, I would check that that number is below INT_MAX, and return some error otherwise (not enough memory or something). If the number is reasonable, we keep doing as today.